docs/CONTRIBUTING.md

```diff
@@ -35,3 +35,47 @@ Test is determined fragile/broken when it is failing despite no defects are pres
   - Test can still be resurrected from history if needed - in ATH or plugin itself.
 
 The situation is expected to be handled in 90 days since reported. In case of no corrective action, the test will be removed from ATH.
+
+## Guidelines to avoid fragile/flaky tests
+
+### Click and wait
+
+When writing acceptance tests using Selenium it is common to click on an element and then run some other action on the resulting page.
+Most of the time the page is updated really quick (specially if there is not a network request involved), but on slow infrastructure this can take some time so that the next test action after clicking could be checking on something which is not the expected.
+
+To avoid this, use the click and wait pattern:
+
+```
+WebElement clickableElement = ...;
+clickableElement.click();
+waitFor(by.id("my-element")).withTimeout(10);
+... now run actions safely on my-element ...
+```
+
+This way the test will not be affected by an underlying slow infrastructure.
+
+### Do not rely on external infra
+
+When the test requires to interact with an external system, use a `DockerFixture`.
+There are multiple examples into the [`fixtures` directory](https://github.com/jenkinsci/acceptance-test-harness/tree/master/src/main/java/org/jenkinsci/test/acceptance/docker/fixtures).
+
+### Do not use sleep
+
+Sleeping blindly is a bad practice, it just makes tests execution time longer. 
+Use the [`Wait` class](https://github.com/jenkinsci/acceptance-test-harness/blob/master/src/main/java/org/jenkinsci/test/acceptance/junit/Wait.java) as much as possible. `CapybaraPortingLayerImpl#waitFor` is an easy way to create instances of `Wait`.
+
+### Use robust selectors
+
+When selecting elements from the DOM, use robust selectors. 
+In general, try to avoid selectors that are known to change or are outside your control (they are not in your code). 
+
+A few examples:
+
+1. Do not use full xpath expressions. DOM can change anytime.
+2. Avoid using parts of the DOM which are not under your control (a change in core or any other plugin can make your test to fail).
+3. Use paths generated by `form-element-path` plugin as much as possible (installed by default by the harness code). Look for the `path` attribute in DOM elements.
+
+### Use existing PageObjects
+
+If your test needs to interact with Jenkins Core (or other plugins UI), check if there is already a PageObject for it instead of writing the code by yourself.
+Existing PageObjects are usually well tested so they are possibly more stable than your newly added code.
```

Comment:
prefer to write a new page object for re-use than to hard code elements etc in your tests?